content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+
+- [Registry Implementation](https://github.com/aenix-io/cozystack/blob/003edf8cf0a419bd67cd822d61ff806db49e7026/pkg/registry/apps/application/rest.go)
+
+At the end of each method, we set the correct Kind and return an `unstructured.Unstructured{}` object
+so that Kubernetes serializes the object correctly. Otherwise,
+it would always serialize them with `kind: Application`, which we don't want.
+
+
+## What did we achieve?
+
+In Cozystack, all our types from the ConfigMap are now available in Kubernetes as-is:
+
+```shell
+# kubectl api-resources | grep cozystack
+buckets                   apps.cozystack.io/v1alpha1      true        Bucket
+clickhouses               apps.cozystack.io/v1alpha1      true        ClickHouse
+etcds                     apps.cozystack.io/v1alpha1      true        Etcd
+ferretdb                  apps.cozystack.io/v1alpha1      true        FerretDB
+httpcaches                apps.cozystack.io/v1alpha1      true        HTTPCache
+ingresses                 apps.cozystack.io/v1alpha1      true        Ingress
+kafkas                    apps.cozystack.io/v1alpha1      true        Kafka
+kuberneteses              apps.cozystack.io/v1alpha1      true        Kubernetes
+monitorings               apps.cozystack.io/v1alpha1      true        Monitoring
+mysqls                    apps.cozystack.io/v1alpha1      true        MySQL
+natses                    apps.cozystack.io/v1alpha1      true        NATS
+postgreses                apps.cozystack.io/v1alpha1      true        Postgres
+rabbitmqs                 apps.cozystack.io/v1alpha1      true        RabbitMQ
+redises                   apps.cozystack.io/v1alpha1      true        Redis
+seaweedfses               apps.cozystack.io/v1alpha1      true        SeaweedFS
+tcpbalancers              apps.cozystack.io/v1alpha1      true        TCPBalancer
+tenants                   apps.cozystack.io/v1alpha1      true        Tenant
+virtualmachines           apps.cozystack.io/v1alpha1      true        VirtualMachine
+vmdisks                   apps.cozystack.io/v1alpha1      true        VMDisk
+vminstances               apps.cozystack.io/v1alpha1      true        VMInstance
+vpns                      apps.cozystack.io/v1alpha1      true        VPN
+```
+
+We can work with them just like regular Kubernetes resources:
+
+```shell
+# kubectl get buckets.apps.cozystack.io -n tenant-kvaps
+NAME         READY   AGE    VERSION
+foo          True    22h    0.1.0
+testaasd     True    27h    0.1.0
+
+# kubectl get kuberneteses.apps.cozystack.io -n tenant-kvaps
+NAME     READY   AGE    VERSION
+abc      False   19h    0.14.0
+asdte    True    22h    0.13.0
+
+# kubectl get vmdisks.apps.cozystack.io -n tenant-kvaps
+NAME               READY   AGE    VERSION
+docker             True    21d    0.1.0
+test               True    18d    0.1.0
+win2k25-iso        True    21d    0.1.0
+win2k25-system     True    21d    0.1.0
+
+# kubectl get vminstances.apps.cozystack.io -n tenant-kvaps
```

Comment:
[Separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output) (and don't include the prompt)